File: bindings/swift/Sources/EzoneSDK/EzoneClient.swift
import Foundation

public struct EzoneError: Error, LocalizedError {
    public let statusCode: Int
    public let message: String
    public var errorDescription: String? { "EzoneError(\(statusCode)): \(message)" }
}

// ── Response models ───────────────────────────────────────────────────────────

public struct SessionResponse: Codable {
    public let token:     String
    public let expiresAt: Int
    public let userId:    String?
    public let deviceId:  String?
    enum CodingKeys: String, CodingKey {
        case token, userId = "user_id", deviceId = "device_id", expiresAt = "expires_at"
    }
}

public struct SessionInfo: Codable {
    public let userId:    String
    public let email:     String
    public let expiresAt: Int
    enum CodingKeys: String, CodingKey {
        case userId = "user_id", email, expiresAt = "expires_at"
    }
}

public struct BeginRegistrationResponse: Codable {
    public let magicToken: String
    enum CodingKeys: String, CodingKey { case magicToken = "magic_token" }
}

public struct BeginLoginResponse: Codable {
    public let challenge: String
}

public struct Device: Codable {
    public let deviceId:     String
    public let deviceName:   String
    public let registeredAt: Int
    enum CodingKeys: String, CodingKey {
        case deviceId = "device_id", deviceName = "device_name", registeredAt = "registered_at"
    }
}

public struct DevicesResponse: Codable {
    public let devices: [Device]
}

public struct RecoveryCodesResponse: Codable {
    public let codes: [String]
}

// ── Client ────────────────────────────────────────────────────────────────────

public class EzoneClient {
    private let baseURL: URL
    private let session: URLSession

    public init(baseURL: URL, timeout: TimeInterval = 10) {
        self.baseURL = baseURL
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        self.session = URLSession(configuration: config)
    }

    // ── Registration ─────────────────────────────────────────────────────────

    public func beginRegistration(email: String) async throws -> BeginRegistrationResponse {
        try await post("/v1/auth/register/begin", body: ["email": email])
    }

    public func completeRegistration(
        magicToken: String, devicePublicKey: String, deviceName: String
    ) async throws -> SessionResponse {
        try await post("/v1/auth/register/complete", body: [
            "magic_token": magicToken, "device_public_key": devicePublicKey, "device_name": deviceName,
        ])
    }

    // ── Login ─────────────────────────────────────────────────────────────────

    public func beginLogin(email: String) async throws -> BeginLoginResponse {
        try await post("/v1/auth/login/begin", body: ["email": email])
    }

    public func completeLogin(
        email: String, challenge: String, signature: String, devicePublicKey: String
    ) async throws -> SessionResponse {
        try await post("/v1/auth/login/complete", body: [
            "email": email, "challenge": challenge,
            "signature": signature, "device_public_key": devicePublicKey,
        ])
    }

    // ── Session ───────────────────────────────────────────────────────────────

    public func verifySession(_ token: String) async throws -> SessionInfo {
        try await get("/v1/auth/session", token: token)
    }

    public func refreshSession(_ token: String) async throws -> SessionResponse {
        try await post("/v1/auth/session/refresh", body: [:], token: token)
    }

    public func logout(_ token: String) async throws {
        let _: [String: Bool] = try await post("/v1/auth/session/logout", body: [:], token: token)
    }

    // ── Reset ─────────────────────────────────────────────────────────────────

    public func beginReset(email: String) async throws -> BeginRegistrationResponse {
        try await post("/v1/auth/reset/begin", body: ["email": email])
    }

    public func completeReset(
        magicToken: String, devicePublicKey: String, deviceName: String
    ) async throws -> SessionResponse {
        try await post("/v1/auth/reset/complete", body: [
            "magic_token": magicToken, "device_public_key": devicePublicKey, "device_name": deviceName,
        ])
    }

    // ── Recovery ──────────────────────────────────────────────────────────────

    public func generateRecoveryCodes(_ token: String) async throws -> RecoveryCodesResponse {
        try await post("/v1/auth/recovery/generate", body: [:], token: token)
    }

    public func recoverWithCode(
        email: String, code: String, devicePublicKey: String, deviceName: String
    ) async throws -> SessionResponse {
        try await post("/v1/auth/recovery/use", body: [
            "email": email, "code": code,
            "device_public_key": devicePublicKey, "device_name": deviceName,
        ])
    }

    // ── Devices ───────────────────────────────────────────────────────────────

    public func listDevices(_ token: String) async throws -> DevicesResponse {
        try await get("/v1/auth/devices", token: token)
    }

    public func beginAddDevice(_ token: String) async throws -> BeginRegistrationResponse {
        try await post("/v1/auth/devices/add/begin", body: [:], token: token)
    }

    public func completeAddDevice(
        magicToken: String, devicePublicKey: String, deviceName: String
    ) async throws -> Device {
        try await post("/v1/auth/devices/add/complete", body: [
            "magic_token": magicToken, "device_public_key": devicePublicKey, "device_name": deviceName,
        ])
    }

    public func revokeDevice(_ token: String, deviceId: String) async throws {
        let url = baseURL.appendingPathComponent("v1/auth/devices/\(deviceId)")
        var req = URLRequest(url: url)
        req.httpMethod = "DELETE"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (_, resp) = try await session.data(for: req)
        guard let http = resp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw EzoneError(statusCode: (resp as? HTTPURLResponse)?.statusCode ?? 0,
                             message: "revoke failed")
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private func post<T: Decodable>(
        _ path: String,
        body: [String: String],
        token: String? = nil
    ) async throws -> T {
        var req = URLRequest(url: baseURL.appendingPathComponent(path.dropFirst()))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let t = token { req.setValue("Bearer \(t)", forHTTPHeaderField: "Authorization") }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await send(req)
    }

    private func get<T: Decodable>(_ path: String, token: String) async throws -> T {
        var req = URLRequest(url: baseURL.appendingPathComponent(path.dropFirst()))
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return try await send(req)
    }

    private func send<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, resp) = try await session.data(for: req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        if status < 200 || status >= 300 {
            let err = try? JSONDecoder().decode([String: String].self, from: data)
            throw EzoneError(statusCode: status, message: err?["error"] ?? "error")
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

File: bindings/java/src/main/java/io/ezone/EzoneClient.java
package io.ezone;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;

public class EzoneClient {

    private final String   baseUrl;
    private final HttpClient http;
    private final ObjectMapper mapper = new ObjectMapper();

    public EzoneClient(String baseUrl) {
        this(baseUrl, Duration.ofSeconds(10));
    }

    public EzoneClient(String baseUrl, Duration timeout) {
        this.baseUrl = baseUrl.stripTrailing();
        this.http    = HttpClient.newBuilder()
                .connectTimeout(timeout)
                .build();
    }

    // ── Registration ─────────────────────────────────────────────────────────

    public Map<String, Object> beginRegistration(String email) throws EzoneException {
        return post("/v1/auth/register/begin", Map.of("email", email), null);
    }

    public Map<String, Object> completeRegistration(
            String magicToken, String devicePublicKey, String deviceName) throws EzoneException {
        return post("/v1/auth/register/complete", Map.of(
                "magic_token",       magicToken,
                "device_public_key", devicePublicKey,
                "device_name",       deviceName), null);
    }

    // ── Login ─────────────────────────────────────────────────────────────────

    public Map<String, Object> beginLogin(String email) throws EzoneException {
        return post("/v1/auth/login/begin", Map.of("email", email), null);
    }

    public Map<String, Object> completeLogin(
            String email, String challenge, String signature, String devicePublicKey)
            throws EzoneException {
        return post("/v1/auth/login/complete", Map.of(
                "email",             email,
                "challenge",         challenge,
                "signature",         signature,
                "device_public_key", devicePublicKey), null);
    }

    // ── Session ───────────────────────────────────────────────────────────────

    public Map<String, Object> verifySession(String token) throws EzoneException {
        return get("/v1/auth/session", token);
    }

    public Map<String, Object> refreshSession(String token) throws EzoneException {
        return post("/v1/auth/session/refresh", Map.of(), token);
    }

    public void logout(String token) throws EzoneException {
        post("/v1/auth/session/logout", Map.of(), token);
    }

    // ── Reset ─────────────────────────────────────────────────────────────────

    public Map<String, Object> beginReset(String email) throws EzoneException {
        return post("/v1/auth/reset/begin", Map.of("email", email), null);
    }

    public Map<String, Object> completeReset(
            String magicToken, String devicePublicKey, String deviceName) throws EzoneException {
        return post("/v1/auth/reset/complete", Map.of(
                "magic_token",       magicToken,
                "device_public_key", devicePublicKey,
                "device_name",       deviceName), null);
    }

    // ── Recovery ──────────────────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    public List<String> generateRecoveryCodes(String token) throws EzoneException {
        Map<String, Object> resp = post("/v1/auth/recovery/generate", Map.of(), token);
        return (List<String>) resp.get("codes");
    }

    public Map<String, Object> recoverWithCode(
            String email, String code, String devicePublicKey, String deviceName)
            throws EzoneException {
        return post("/v1/auth/recovery/use", Map.of(
                "email",             email,
                "code",              code,
                "device_public_key", devicePublicKey,
                "device_name",       deviceName), null);
    }

    // ── Devices ───────────────────────────────────────────────────────────────

    public Map<String, Object> listDevices(String token) throws EzoneException {
        return get("/v1/auth/devices", token);
    }

    public Map<String, Object> beginAddDevice(String token) throws EzoneException {
        return post("/v1/auth/devices/add/begin", Map.of(), token);
    }

    public Map<String, Object> completeAddDevice(
            String magicToken, String devicePublicKey, String deviceName) throws EzoneException {
        return post("/v1/auth/devices/add/complete", Map.of(
                "magic_token",       magicToken,
                "device_public_key", devicePublicKey,
                "device_name",       deviceName), null);
    }

    public void revokeDevice(String token, String deviceId) throws EzoneException {
        delete("/v1/auth/devices/" + deviceId, token);
    }

    // ── HTTP helpers ──────────────────────────────────────────────────────────

    @SuppressWarnings("unchecked")
    private Map<String, Object> post(String path, Object body, String token)
            throws EzoneException {
        try {
            var reqBuilder = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body)));
            if (token != null) reqBuilder.header("Authorization", "Bearer " + token);
            var resp = http.send(reqBuilder.build(), HttpResponse.BodyHandlers.ofString());
            return parseResponse(resp);
        } catch (IOException | InterruptedException e) {
            throw new EzoneException(0, e.getMessage());
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> get(String path, String token) throws EzoneException {
        try {
            var req = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Authorization", "Bearer " + token)
                    .GET()
                    .build();
            var resp = http.send(req, HttpResponse.BodyHandlers.ofString());
            return parseResponse(resp);
        } catch (IOException | InterruptedException e) {
            throw new EzoneException(0, e.getMessage());
        }
    }

    private void delete(String path, String token) throws EzoneException {
        try {
            var req = HttpRequest.newBuilder()
                    .uri(URI.create(baseUrl + path))
                    .header("Authorization", "Bearer " + token)
                    .DELETE()
                    .build();
            var resp = http.send(req, HttpResponse.BodyHandlers.ofString());
            if (resp.statusCode() < 200 || resp.statusCode() >= 300) {
                var body = mapper.readValue(resp.body(), Map.class);
                throw new EzoneException(resp.statusCode(),
                        (String) body.getOrDefault("error", "error"));
            }
        } catch (IOException | InterruptedException e) {
            throw new EzoneException(0, e.getMessage());
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseResponse(HttpResponse<String> resp) throws EzoneException {
        try {
            var body = mapper.readValue(resp.body(), Map.class);
            if (resp.statusCode() < 200 || resp.statusCode() >= 300) {
                throw new EzoneException(resp.statusCode(),
                        (String) body.getOrDefault("error", "error"));
            }
            return body;
        } catch (IOException e) {
            throw new EzoneException(resp.statusCode(), "failed to parse response");
        }
    }
}
